synapse/handlers/identity.py

```diff
@@ -408,33 +404,33 @@ def send_threepid_validation(
 
         return session_id
 
-    def rewrite_id_server_url(self, url: str) -> str:
-        """Given an identity server URL, rewrite it according to the
-        rewrite_identity_server_urls config option
+    def rewrite_id_server_url(self, url: str, add_https=False) -> str:
+        """Given an identity server URL, optionally add a protocol scheme
+        before rewriting it according to the rewrite_identity_server_urls
+        config option
 
-        First removes the protocol scheme from the URL if provided.
-        Then checks for a rewritten URL. If found, returns the new URL.
-        Otherwise, returns the original URL.
+        Adds https:// to the URL if specified, then tries to rewrite the
+        url. Returns either the rewritten URL or the URL with optional
+        protocol scheme additions.
         """
         rewritten_url = url
-        if url.startswith("http://"):
-            rewritten_url = url[7:]
-        elif url.startswith("https://"):
-            rewritten_url = url[8:]
-
-        return self.rewrite_identity_server_urls.get(rewritten_url, url)
+        if add_https:
```

Comment:
We should probably also check here that the URL doesn't already begins with `https://` (or `http://`), otherwise we'll end up with an invalid entry.

Reply:
I'm a little reluctant to as that means a client is sending the wrong thing, and we may end up in a situation where a client is sending its `id_server` params as `https://blabla.com` which Synapse happily accepts, but another server implementation does not.

That said, Synapse probably won't fail in the most graceful here.

Reply:
Yeah I think we should be as flexible as possible here, and allow for client mishaps as long as it doesn't require too much work (in this case it barely costs anything).

Reply:
Having experience from https://github.com/matrix-org/synapse/issues/6766 it can be a lot of work to inform client developers that they're doing the wrong thing.

Reply:
Fair enough then.